In [10]:
import pandas as pd
import numpy as np
np.random.seed(10)

# Reduce the number of rows by random sampling

In order to reduce the number of rows we are selecting random indexs of the original data frame and deleting them.

In [11]:
#reducing the number of rows to 10000
df = pd.read_csv(r'../RawDatasets/bank-additional-full.csv',  sep=';', na_values=".")
nRowsToRemove = len(df) - 10000
rowsToDropIndices = np.random.choice(df.index, nRowsToRemove, replace = False)
df_reduced = df.drop(rowsToDropIndices)
df_reduced.to_csv(r'../Datasets/bank-additional-reduced.csv')

# Study if the result is representative
After doing a sample of the original data set we need to test that the result is representative and that for each column we have a similar statistical structure to
the original dataset.

## Description of all the features 
With these descriptions we can already observe for the numerical variables if the mean, std, min and max values or others have changed or not. If the values are very similar, we can accept the previous algorithm to reduce the number of rows and proceed with the project.

In [12]:
#study of the  proportions/frequencies of each feature in the original and row-reduced data set
df.describe(include='all')

age     job  marital          education default housing  \
count   41188.00000   41188    41188              41188   41188   41188   
unique          NaN      12        4                  8       3       3   
top             NaN  admin.  married  university.degree      no     yes   
freq            NaN   10422    24928              12168   32588   21576   
mean       40.02406     NaN      NaN                NaN     NaN     NaN   
std        10.42125     NaN      NaN                NaN     NaN     NaN   
min        17.00000     NaN      NaN                NaN     NaN     NaN   
25%        32.00000     NaN      NaN                NaN     NaN     NaN   
50%        38.00000     NaN      NaN                NaN     NaN     NaN   
75%        47.00000     NaN      NaN                NaN     NaN     NaN   
max        98.00000     NaN      NaN                NaN     NaN     NaN   

         loan   contact  month day_of_week  ...      campaign         pdays  \
count   41188     41188  41188       41188  ...  41188.000000  41188.000000   
unique      3         2     10           5  ...           NaN           NaN   
top        no  cellular    may         thu  ...           NaN           NaN   
freq    33950     26144  13769        8623  ...           NaN           NaN   
mean      NaN       NaN    NaN         NaN  ...      2.567593    962.475454   
std       NaN       NaN    NaN         NaN  ...      2.770014    186.910907   
min       NaN       NaN    NaN         NaN  ...      1.000000      0.000000   
25%       NaN       NaN    NaN         NaN  ...      1.000000    999.000000   
50%       NaN       NaN    NaN         NaN  ...      2.000000    999.000000   
75%       NaN       NaN    NaN         NaN  ...      3.000000    999.000000   
max       NaN       NaN    NaN         NaN  ...     56.000000    999.000000   

            previous     poutcome  emp.var.rate  cons.price.idx  \
count   41188.000000        41188  41188.000000    41188.000000   
unique           NaN            3           NaN             NaN   
top              NaN  nonexistent           NaN             NaN   
freq             NaN        35563           NaN             NaN   
mean        0.172963          NaN      0.081886       93.575664   
std         0.494901          NaN      1.570960        0.578840   
min         0.000000          NaN     -3.400000       92.201000   
25%         0.000000          NaN     -1.800000       93.075000   
50%         0.000000          NaN      1.100000       93.749000   
75%         0.000000          NaN      1.400000       93.994000   
max         7.000000          NaN      1.400000       94.767000   

        cons.conf.idx     euribor3m   nr.employed      y  
count    41188.000000  41188.000000  41188.000000  41188  
unique            NaN           NaN           NaN      2  
top               NaN           NaN           NaN     no  
freq              NaN           NaN           NaN  36548  
mean       -40.502600      3.621291   5167.035911    NaN  
std          4.628198      1.734447     72.251528    NaN  
min        -50.800000      0.634000   4963.600000    NaN  
25%        -42.700000      1.344000   5099.100000    NaN  
50%        -41.800000      4.857000   5191.000000    NaN  
75%        -36.400000      4.961000   5228.100000    NaN  
max        -26.900000      5.045000   5228.100000    NaN  

[11 rows x 21 columns]

In [13]:
df_reduced.describe(include='all')

age     job  marital          education default housing  \
count   10000.000000   10000    10000              10000   10000   10000   
unique           NaN      12        4                  8       3       3   
top              NaN  admin.  married  university.degree      no     yes   
freq             NaN    2534     6029               2931    7888    5196   
mean       39.919700     NaN      NaN                NaN     NaN     NaN   
std        10.331211     NaN      NaN                NaN     NaN     NaN   
min        17.000000     NaN      NaN                NaN     NaN     NaN   
25%        32.000000     NaN      NaN                NaN     NaN     NaN   
50%        38.000000     NaN      NaN                NaN     NaN     NaN   
75%        47.000000     NaN      NaN                NaN     NaN     NaN   
max        91.000000     NaN      NaN                NaN     NaN     NaN   

         loan   contact  month day_of_week  ...      campaign        pdays  \
count   10000     10000  10000       10000  ...  10000.000000  10000.00000   
unique      3         2     10           5  ...           NaN          NaN   
top        no  cellular    may         thu  ...           NaN          NaN   
freq     8247      6381   3363        2119  ...           NaN          NaN   
mean      NaN       NaN    NaN         NaN  ...      2.594000    961.26540   
std       NaN       NaN    NaN         NaN  ...      2.834955    189.87182   
min       NaN       NaN    NaN         NaN  ...      1.000000      0.00000   
25%       NaN       NaN    NaN         NaN  ...      1.000000    999.00000   
50%       NaN       NaN    NaN         NaN  ...      2.000000    999.00000   
75%       NaN       NaN    NaN         NaN  ...      3.000000    999.00000   
max       NaN       NaN    NaN         NaN  ...     56.000000    999.00000   

            previous     poutcome  emp.var.rate  cons.price.idx  \
count   10000.000000        10000  10000.000000    10000.000000   
unique           NaN            3           NaN             NaN   
top              NaN  nonexistent           NaN             NaN   
freq             NaN         8615           NaN             NaN   
mean        0.177900          NaN      0.053900       93.570333   
std         0.509782          NaN      1.579857        0.581010   
min         0.000000          NaN     -3.400000       92.201000   
25%         0.000000          NaN     -1.800000       93.075000   
50%         0.000000          NaN      1.100000       93.749000   
75%         0.000000          NaN      1.400000       93.994000   
max         6.000000          NaN      1.400000       94.767000   

        cons.conf.idx     euribor3m   nr.employed      y  
count    10000.000000  10000.000000  10000.000000  10000  
unique            NaN           NaN           NaN      2  
top               NaN           NaN           NaN     no  
freq              NaN           NaN           NaN   8845  
mean       -40.570390      3.585578   5165.684620    NaN  
std          4.662535      1.747843     72.774978    NaN  
min        -50.800000      0.634000   4963.600000    NaN  
25%        -42.700000      1.334000   5099.100000    NaN  
50%        -41.800000      4.857000   5191.000000    NaN  
75%        -36.400000      4.961000   5228.100000    NaN  
max        -26.900000      5.045000   5228.100000    NaN  

[11 rows x 21 columns]

To validate that the categorical variables values have the same representation in both the original and the reduced data sets is not enough to observe the previous description. We must observe that the percentage of each feature value has not changed much.

In [14]:
print ("Job feature percentatges in the original data set")
print(df.job.value_counts(normalize=True))
print ("Job feature percentatges in the reduced data set")
print(df_reduced.job.value_counts(normalize=True))

Job feature percentatges in the original data set
admin.           0.253035
blue-collar      0.224677
technician       0.163713
services         0.096363
management       0.070992
retired          0.041760
entrepreneur     0.035350
self-employed    0.034500
housemaid        0.025736
unemployed       0.024619
student          0.021244
unknown          0.008012
Name: job, dtype: float64
Job feature percentatges in the reduced data set
admin.           0.2534
blue-collar      0.2335
technician       0.1564
services         0.0964
management       0.0698
retired          0.0416
self-employed    0.0359
entrepreneur     0.0353
housemaid        0.0251
unemployed       0.0236
student          0.0217
unknown          0.0073
Name: job, dtype: float64


In [15]:
print("Difference between the percentage of each 'job' value in the original data set and the reduced data set")
print(df.job.value_counts(normalize=True)-df_reduced.job.value_counts(normalize=True))

Difference between the percentage of each 'job' value in the original data set and the reduced data set
admin.          -0.000365
blue-collar     -0.008823
entrepreneur     0.000050
housemaid        0.000636
management       0.001192
retired          0.000160
self-employed   -0.001400
services        -0.000037
student         -0.000456
technician       0.007313
unemployed       0.001019
unknown          0.000712
Name: job, dtype: float64


In [16]:
print("Difference between the percentage of each 'marital' value in the original data set and the reduced data set")
print(df.marital.value_counts(normalize=True)-df_reduced.marital.value_counts(normalize=True))

Difference between the percentage of each 'marital' value in the original data set and the reduced data set
married     0.002325
single     -0.001641
divorced   -0.000726
unknown     0.000042
Name: marital, dtype: float64


In [17]:
print("Difference between the percentage of each 'education' value in the original data set and the reduced data set")
print(df.education.value_counts(normalize=True)-df_reduced.education.value_counts(normalize=True))

Difference between the percentage of each 'education' value in the original data set and the reduced data set
university.degree      0.002326
high.school            0.003414
basic.9y              -0.004134
professional.course    0.004294
basic.4y              -0.000911
basic.6y              -0.004153
unknown               -0.000873
illiterate             0.000037
Name: education, dtype: float64


In [18]:
print("Difference between the percentage of each 'default' value in the original data set and the reduced data set")
print(df.default.value_counts(normalize=True)-df_reduced.default.value_counts(normalize=True))

Difference between the percentage of each 'default' value in the original data set and the reduced data set
no         0.002401
unknown   -0.002374
yes       -0.000027
Name: default, dtype: float64


In [19]:
print("Difference between the percentage of each 'housing' value in the original data set and the reduced data set")
print(df.housing.value_counts(normalize=True)-df_reduced.housing.value_counts(normalize=True))

Difference between the percentage of each 'housing' value in the original data set and the reduced data set
yes        0.004242
no        -0.002478
unknown   -0.001764
Name: housing, dtype: float64


In [20]:
print("Difference between the percentage of each 'loan' value in the original data set and the reduced data set")
print(df.loan.value_counts(normalize=True)-df_reduced.loan.value_counts(normalize=True))

Difference between the percentage of each 'loan' value in the original data set and the reduced data set
no        -0.000431
yes        0.002195
unknown   -0.001764
Name: loan, dtype: float64


In [21]:
print("Difference between the percentage of each 'contact' value in the original data set and the reduced data set")
print(df.contact.value_counts(normalize=True)-df_reduced.contact.value_counts(normalize=True))

Difference between the percentage of each 'contact' value in the original data set and the reduced data set
cellular    -0.003352
telephone    0.003352
Name: contact, dtype: float64


In [22]:
print("Difference between the percentage of each 'month' value in the original data set and the reduced data set")
print(df.month.value_counts(normalize=True)-df_reduced.month.value_counts(normalize=True))

Difference between the percentage of each 'month' value in the original data set and the reduced data set
apr   -0.002798
aug    0.004195
dec    0.000219
jul   -0.003923
jun    0.000615
mar   -0.000744
may   -0.002004
nov    0.004368
oct   -0.001068
sep    0.001139
Name: month, dtype: float64


In [23]:
print("Difference between the percentage of each 'day_of_week' value in the original data set and the reduced data set")
print(df.day_of_week.value_counts(normalize=True)-df_reduced.day_of_week.value_counts(normalize=True))

Difference between the percentage of each 'day_of_week' value in the original data set and the reduced data set
fri    0.003331
mon   -0.002089
thu   -0.002543
tue   -0.005584
wed    0.006885
Name: day_of_week, dtype: float64


In [24]:
print("Difference between the percentage of each 'poutcome' value in the original data set and the reduced data set")
print(df.poutcome.value_counts(normalize=True)-df_reduced.poutcome.value_counts(normalize=True))

Difference between the percentage of each 'poutcome' value in the original data set and the reduced data set
nonexistent    0.001931
failure       -0.000966
success       -0.000965
Name: poutcome, dtype: float64


In [25]:
print("Difference between the percentage of each 'y' value in the original data set and the reduced data set")
print(df.y.value_counts(normalize=True)-df_reduced.y.value_counts(normalize=True))

Difference between the percentage of each 'y' value in the original data set and the reduced data set
no     0.002846
yes   -0.002846
Name: y, dtype: float64


We can conclude that the sample obtained is representative and we can use it for our project.

# Reduce the number of rows to balance the classes
The previous reduction of rows does not take into account if the dataset is balanced or not. As our data set is not balanced, we  can reduce the data set by eliminating only rows of the class with bigger proportion.

In [26]:
df.y.value_counts(normalize=True)

no     0.887346
yes    0.112654
Name: y, dtype: float64

As it can be observed, the rows with y=no represent the 88.7% (36548 rows) of the totality and rows with y=yes represent only the 11.26% (4640). Our data set is very  unbalanced so, to solve this problem, we can eliminate rows that have y=no until having 10k in total.

In [27]:
#reducing the number of rows to 10000
nRowsToRemove = len(df) - 10000 #we could higher the number of rows to remove to end with exactly the same proportion of yes and no (9280 rows in total)
rowsToDropIndices = np.random.choice(df[df.y == 'no'].index, nRowsToRemove, replace = False)
df_balanced = df.drop(rowsToDropIndices)
df_balanced.to_csv(r'../Datasets/bank-additional-reduced-balanced.csv')

In [28]:
df_balanced.y.value_counts(normalize=True)

no     0.536
yes    0.464
Name: y, dtype: float64

In [29]:
#study of the  proportions/frequencies of each feature in the original and row-reduced data set
df.describe(include='all')

age     job  marital          education default housing  \
count   41188.00000   41188    41188              41188   41188   41188   
unique          NaN      12        4                  8       3       3   
top             NaN  admin.  married  university.degree      no     yes   
freq            NaN   10422    24928              12168   32588   21576   
mean       40.02406     NaN      NaN                NaN     NaN     NaN   
std        10.42125     NaN      NaN                NaN     NaN     NaN   
min        17.00000     NaN      NaN                NaN     NaN     NaN   
25%        32.00000     NaN      NaN                NaN     NaN     NaN   
50%        38.00000     NaN      NaN                NaN     NaN     NaN   
75%        47.00000     NaN      NaN                NaN     NaN     NaN   
max        98.00000     NaN      NaN                NaN     NaN     NaN   

         loan   contact  month day_of_week  ...      campaign         pdays  \
count   41188     41188  41188       41188  ...  41188.000000  41188.000000   
unique      3         2     10           5  ...           NaN           NaN   
top        no  cellular    may         thu  ...           NaN           NaN   
freq    33950     26144  13769        8623  ...           NaN           NaN   
mean      NaN       NaN    NaN         NaN  ...      2.567593    962.475454   
std       NaN       NaN    NaN         NaN  ...      2.770014    186.910907   
min       NaN       NaN    NaN         NaN  ...      1.000000      0.000000   
25%       NaN       NaN    NaN         NaN  ...      1.000000    999.000000   
50%       NaN       NaN    NaN         NaN  ...      2.000000    999.000000   
75%       NaN       NaN    NaN         NaN  ...      3.000000    999.000000   
max       NaN       NaN    NaN         NaN  ...     56.000000    999.000000   

            previous     poutcome  emp.var.rate  cons.price.idx  \
count   41188.000000        41188  41188.000000    41188.000000   
unique           NaN            3           NaN             NaN   
top              NaN  nonexistent           NaN             NaN   
freq             NaN        35563           NaN             NaN   
mean        0.172963          NaN      0.081886       93.575664   
std         0.494901          NaN      1.570960        0.578840   
min         0.000000          NaN     -3.400000       92.201000   
25%         0.000000          NaN     -1.800000       93.075000   
50%         0.000000          NaN      1.100000       93.749000   
75%         0.000000          NaN      1.400000       93.994000   
max         7.000000          NaN      1.400000       94.767000   

        cons.conf.idx     euribor3m   nr.employed      y  
count    41188.000000  41188.000000  41188.000000  41188  
unique            NaN           NaN           NaN      2  
top               NaN           NaN           NaN     no  
freq              NaN           NaN           NaN  36548  
mean       -40.502600      3.621291   5167.035911    NaN  
std          4.628198      1.734447     72.251528    NaN  
min        -50.800000      0.634000   4963.600000    NaN  
25%        -42.700000      1.344000   5099.100000    NaN  
50%        -41.800000      4.857000   5191.000000    NaN  
75%        -36.400000      4.961000   5228.100000    NaN  
max        -26.900000      5.045000   5228.100000    NaN  

[11 rows x 21 columns]

In [30]:
#study of the  proportions/frequencies of each feature in the original and row-reduced data set
df_balanced.describe(include='all')

age     job  marital          education default housing  \
count   10000.000000   10000    10000              10000   10000   10000   
unique           NaN      12        4                  8       2       3   
top              NaN  admin.  married  university.degree      no     yes   
freq             NaN    2702     5778               3252    8358    5279   
mean       40.214000     NaN      NaN                NaN     NaN     NaN   
std        11.899419     NaN      NaN                NaN     NaN     NaN   
min        17.000000     NaN      NaN                NaN     NaN     NaN   
25%        31.000000     NaN      NaN                NaN     NaN     NaN   
50%        38.000000     NaN      NaN                NaN     NaN     NaN   
75%        48.000000     NaN      NaN                NaN     NaN     NaN   
max        98.000000     NaN      NaN                NaN     NaN     NaN   

         loan   contact  month day_of_week  ...      campaign         pdays  \
count   10000     10000  10000       10000  ...  10000.000000  10000.000000   
unique      3         2     10           5  ...           NaN           NaN   
top        no  cellular    may         thu  ...           NaN           NaN   
freq     8287      7102   2797        2183  ...           NaN           NaN   
mean      NaN       NaN    NaN         NaN  ...      2.382300    894.133100   
std       NaN       NaN    NaN         NaN  ...      2.467013    305.209127   
min       NaN       NaN    NaN         NaN  ...      1.000000      0.000000   
25%       NaN       NaN    NaN         NaN  ...      1.000000    999.000000   
50%       NaN       NaN    NaN         NaN  ...      2.000000    999.000000   
75%       NaN       NaN    NaN         NaN  ...      3.000000    999.000000   
max       NaN       NaN    NaN         NaN  ...     39.000000    999.000000   

            previous     poutcome  emp.var.rate  cons.price.idx  \
count   10000.000000        10000  10000.000000    10000.000000   
unique           NaN            3           NaN             NaN   
top              NaN  nonexistent           NaN             NaN   
freq             NaN         7867           NaN             NaN   
mean        0.303600          NaN     -0.441990       93.490151   
std         0.687222          NaN      1.716586        0.629785   
min         0.000000          NaN     -3.400000       92.201000   
25%         0.000000          NaN     -1.800000       92.893000   
50%         0.000000          NaN     -0.100000       93.444000   
75%         0.000000          NaN      1.400000       93.994000   
max         7.000000          NaN      1.400000       94.767000   

        cons.conf.idx     euribor3m  nr.employed      y  
count    10000.000000  10000.000000  10000.00000  10000  
unique            NaN           NaN          NaN      2  
top               NaN           NaN          NaN     no  
freq              NaN           NaN          NaN   5360  
mean       -40.238570      3.021937   5138.22453    NaN  
std          5.297955      1.887082     86.36719    NaN  
min        -50.800000      0.634000   4963.60000    NaN  
25%        -42.700000      1.250000   5076.20000    NaN  
50%        -41.800000      4.076000   5191.00000    NaN  
75%        -36.400000      4.959000   5228.10000    NaN  
max        -26.900000      5.045000   5228.10000    NaN  

[11 rows x 21 columns]

In [31]:
print("Difference between the percentage of each 'job' value in the original data set and the balanced data set")
print(df.job.value_counts(normalize=True)-df_balanced.job.value_counts(normalize=True))

Difference between the percentage of each 'job' value in the original data set and the balanced data set
admin.          -0.017165
blue-collar      0.033077
entrepreneur     0.004350
housemaid        0.002236
management       0.002192
retired         -0.019740
self-employed    0.001200
services         0.010563
student         -0.016256
technician       0.002013
unemployed      -0.002581
unknown          0.000112
Name: job, dtype: float64


In [32]:
print("Difference between the percentage of each 'marital' value in the original data set and the balanced data set")
print(df.marital.value_counts(normalize=True)-df_balanced.marital.value_counts(normalize=True))

Difference between the percentage of each 'marital' value in the original data set and the balanced data set
married     0.027425
single     -0.029841
divorced    0.002874
unknown    -0.000458
Name: marital, dtype: float64


In [33]:
print("Difference between the percentage of each 'education' value in the original data set and the balanced data set")
print(df.education.value_counts(normalize=True)-df_balanced.education.value_counts(normalize=True))

Difference between the percentage of each 'education' value in the original data set and the balanced data set
university.degree     -0.029774
high.school            0.003114
basic.9y               0.012766
professional.course    0.003394
basic.4y               0.006089
basic.6y               0.007847
unknown               -0.003373
illiterate            -0.000063
Name: education, dtype: float64


In [34]:
print("Difference between the percentage of each 'default' value in the original data set and the balanced data set")
print(df.default.value_counts(normalize=True)-df_balanced.default.value_counts(normalize=True))

Difference between the percentage of each 'default' value in the original data set and the balanced data set
no        -0.044599
unknown    0.044526
yes             NaN
Name: default, dtype: float64


In [35]:
print("Difference between the percentage of each 'housing' value in the original data set and the balanced data set")
print(df.housing.value_counts(normalize=True)-df_balanced.housing.value_counts(normalize=True))

Difference between the percentage of each 'housing' value in the original data set and the balanced data set
yes       -0.004058
no         0.001822
unknown    0.002236
Name: housing, dtype: float64


In [36]:
print("Difference between the percentage of each 'loan' value in the original data set and the balanced data set")
print(df.loan.value_counts(normalize=True)-df_balanced.loan.value_counts(normalize=True))

Difference between the percentage of each 'loan' value in the original data set and the balanced data set
no        -0.004431
yes        0.002195
unknown    0.002236
Name: loan, dtype: float64


In [37]:
print("Difference between the percentage of each 'contact' value in the original data set and the balanced data set")
print(df.contact.value_counts(normalize=True)-df_balanced.contact.value_counts(normalize=True))

Difference between the percentage of each 'contact' value in the original data set and the balanced data set
cellular    -0.075452
telephone    0.075452
Name: contact, dtype: float64


In [38]:
print("Difference between the percentage of each 'month' value in the original data set and the balanced data set")
print(df.month.value_counts(normalize=True)-df_balanced.month.value_counts(normalize=True))

Difference between the percentage of each 'month' value in the original data set and the balanced data set
apr   -0.019098
aug    0.006895
dec   -0.005681
jul    0.010777
jun    0.003615
mar   -0.018344
may    0.054596
nov    0.003868
oct   -0.021368
sep   -0.015261
Name: month, dtype: float64


In [39]:
print("Difference between the percentage of each 'day_of_week' value in the original data set and the balanced data set")
print(df.day_of_week.value_counts(normalize=True)-df_balanced.day_of_week.value_counts(normalize=True))

Difference between the percentage of each 'day_of_week' value in the original data set and the balanced data set
fri   -0.000369
mon    0.011511
thu   -0.008943
tue   -0.001484
wed   -0.000715
Name: day_of_week, dtype: float64


In [40]:
print("Difference between the percentage of each 'poutcome' value in the original data set and the balanced data set")
print(df.poutcome.value_counts(normalize=True)-df_balanced.poutcome.value_counts(normalize=True))

Difference between the percentage of each 'poutcome' value in the original data set and the balanced data set
nonexistent    0.076731
failure       -0.012866
success       -0.063865
Name: poutcome, dtype: float64


In this solution now statistics like the means, std, frequencies and other, have changed a bit more than they did in the previous form of reducing. the number of rows studied. One problem is that for the feature "default" we are losing the modality 'yes'. This modality in the original dataset had pretty low representation, exactly only 3 rows had default = 'yes' value.
Despite that, all the other statistical changes are not that drastic.

Although we have now a balanced dataset, we should consider other ways to balance the data set and pick that one that is a better sample and represents better the original data set.
